File: Non-procedural Programming/Solved/antidvorak/male/orez.pl
vstup(T) :- T = t(t(t(nil,5,nil),10,t(nil,15,nil)),20,t(nil,50,t(nil,100,nil))).

%http://forum.matfyz.info/viewtopic.php?f=169&t=10036&p=39223&hilit=orez#p39223

orez(_, _, nil, nil) :- !.
orez(L, U, t(LS, H, PS), t(NL, H, NP)) :-
    H >= L,
    H =< U,
    orez(L, U, LS, NL),
    orez(L, U, PS, NP), !.

orez(L, U, t(_, H, PS), NP) :-
    H < L,
    orez(L,U,PS, NP), !.

orez(L, U, t(LS, H, _), NL) :-
    H > U,
    orez(L,U,LS, NL), !.

File: Non-procedural Programming/Solved/antihric/male/orez.hs
data BVS a = Nil | Node (BVS a) a (BVS a) deriving (Show, Eq)

--http://forum.matfyz.info/viewtopic.php?f=169&t=8308

testStrom = Node (Node (Node (Nil) 10 (Nil)) 25 (Node (Node (Nil) 30 (Nil)) 35 (Node (Nil) 37 (Nil)))) 50 (Node (Nil) 75 (Node (Node (Nil) 80 (Nil)) 100 (Nil)))

orez :: Ord a => a -> a -> BVS a -> BVS a
orez _ _ Nil = Nil
orez x y (Node l v r) | x <= v && v <= y        = Node (orez x y l) v (orez x y r)
                      | v < x                   = orez x y r
                      | v > y                   = orez x y l